mypy/checkexpr.py

```diff
@@ -4025,82 +3985,153 @@ def lookup_definer(typ: Instance, attr_name: str) -> str | None:
+            != self._lookup_definer(right_type, rev_op_name)
         ):
             # When we do "A() + B()" where B is a subclass of A, we'll actually try calling
-            # B's __radd__ method first, but ONLY if B explicitly defines or overrides the
-            # __radd__ method.
+            # B's __radd__ method first, but ONLY if B explicitly defines or overrides it.
             #
-            # This mechanism lets subclasses "refine" the expected outcome of the operation, even
-            # if they're located on the RHS.
+            # This mechanism lets subclasses "refine" the expected outcome of the operation,
+            # even if they're located on the RHS.
             #
             # As a special case, the alt_promote check makes sure that we don't use the
             # __radd__ method of int if the LHS is a native int type.
-
-            variants_raw = [
+            return [
                 (rev_op_name, right_op, right_type, left_expr),
                 (op_name, left_op, left_type, right_expr),
             ]
-        else:
-            # In all other cases, we do the usual thing and call __add__ first and
-            # __radd__ second when doing "A() + B()".
-
-            variants_raw = [
-                (op_name, left_op, left_type, right_expr),
-                (rev_op_name, right_op, right_type, left_expr),
-            ]
-
-        # STEP 3:
-        # We now filter out all non-existent operators. The 'variants' list contains
-        # all operator methods that are actually present, in the order that Python
-        # attempts to invoke them.
 
-        variants = [(na, op, obj, arg) for (na, op, obj, arg) in variants_raw if op is not None]
+        # In all other cases, we do the usual thing and call __add__ first and
+        # __radd__ second when doing "A() + B()".
+        return [
+            (op_name, left_op, left_type, right_expr),
+            (rev_op_name, right_op, right_type, left_expr),
+        ]
 
-        # STEP 4:
-        # We now try invoking each one. If an operation succeeds, end early and return
-        # the corresponding result. Otherwise, return the result and errors associated
-        # with the first entry.
+    def _attempt_operator_applications(
+        self,
+        op_name: str,
+        variants: list[tuple[str, Type, Type, Expression]],
+        left_type: Type,
+        right_type: Type,
+        left_expr: Expression,
+        right_expr: Expression,
+        context: Context,
+    ) -> tuple[Type, Type]:
+        """Try applying the operator methods and handle possible fallbacks."""
 
         errors = []
         results = []
+
         for name, method, obj, arg in variants:
             with self.msg.filter_errors(save_filtered_errors=True) as local_errors:
-                result = self.check_method_call(name, obj, method, [arg], [ARG_POS], context)
+                result = self.check_method_call(op_name, obj, method, [arg], [ARG_POS], context)
```

Comment:
Please wait until #18995 is merged - this will be a conflict (that PR fixes a bug, `op_name` shouldn't be used here)

Reply:
Ough, it is already merged, sorry. Then please don't introduce that bug back!